Tooltip event rewrite

diesieben07 opened the pull request:
Fixes #8037. Fixes #7995. Fixes #8124.

Currently the Post events are removed, because I do not think they make sense with the changes to the vanilla system.

diesieben07 committed:
cherry pick custom tooltip components

diesieben07 commented:
I've added a pluggable system for custom ClientTooltipComponents as well, hooking into the hardcoded bundle tooltip code.

diesieben07 committed:
Add test mod for custom item tooltip components

diesieben07 committed:
Add test mod methods for GatherComponents and Color event

diesieben07 commented:
Why does the RenderTooltipEvent.Pre allow changing the screen size? How is that helpful? If that functionality is removed and the changing of the font renderer is merged with the Colors event then we only need that one event, not both.

Crimix commented:
Is it possible this could fix the styling issue with wrapped tooltips from #7995?

diesieben07 commented:
Unlikely, as the same underlying method is used. But I can look into that, if line wrapping is reimplemented.

diesieben07 committed:
tooltip event cleanup, try no2

diesieben07 committed:
Custom tooltip work

diesieben07 committed:
Reenable Pre and Color events

diesieben07 committed:
add max width test

diesieben07 committed:
Documentation, deprecate old events, patch simplifications, implement cancellation for GatherComponents

diesieben07 commented:
- GatherComponents event implement with ability to set maximum width
- Reimplement text splitting
- Pre event reimplemented with removed ability to set maximum width 
- Color event reimplemented

diesieben07 committed:
Add test for tooltip splitting

diesieben07 commented:
@Crimix Consider yourself lucky, #7995 is fixed by this now:
![2021-08-30_23 41 30](https://user-images.githubusercontent.com/1915984/131409531-44bfa7f1-7c37-4f4a-a267-839fc092aa91.png)

diesieben07 committed:
Deprecate duplicated GuiUtils hover code
Add method to draw tooltip with font, fix Creative Screen patch

diesieben07 committed:
Fix code formatting

Crimix commented:
Nice, good job @diesieben07 thank you 😄

JTK222 commented:
I'd appreciate if it would be possible to add non text components, here's an example usage:
![image](https://user-images.githubusercontent.com/10399186/131448856-e11a03eb-e3bc-461c-a47d-b20388258d94.png)

Possibly just a render component that allows you to specify a size which is considered when rendering the BG?

diesieben07 commented:
@JTK222 A pluggable system for that is included with this PR, check the example mod.

JTK222 commented:
Oh, perfect sorry I assumed this only works with Text Component, then this entire thing looks perfect from my point of view ^^

Mysterious-Dev commented:
Look good

pupnewfster commented on patches/minecraft/net/minecraft/client/gui/screens/Screen.java.patch:
```diff
@@ -1,68 +1,104 @@
-+      this.renderTooltipInternal(p_96618_, p_96619_.stream().map(ClientTooltipComponent::m_169948_).collect(Collectors.toList()), p_96620_, p_96621_, font);
+       this.m_169383_(p_96618_, p_96619_.stream().map(ClientTooltipComponent::m_169948_).collect(Collectors.toList()), p_96620_, p_96621_);
     }
++   public void renderToolTip(PoseStack p_96618_, List<? extends FormattedCharSequence> p_96619_, int p_96620_, int p_96621_, Font font) {
```
This method should have proper param names instead of srg param names, as the entire body of it is forge added. Also this should probably be named `renderTooltip` to have consistent capitalization in the word `Tooltip` with all the other tooltip methods.

pupnewfster commented on patches/minecraft/net/minecraft/client/gui/screens/Screen.java.patch:
```diff
@@ -1,68 +1,104 @@
++      tooltipStack = ItemStack.f_41583_;
     }
  
++   public void renderTooltip(PoseStack p_169389_, List<Component> p_169390_, Optional<TooltipComponent> p_169391_, int p_169392_, int p_169393_, Font font) {
```
This method should have proper param names instead of srg param names, as the entire body of it is forge added.

pupnewfster commented on patches/minecraft/net/minecraft/client/gui/screens/Screen.java.patch:
```diff
@@ -1,68 +1,104 @@
     public void m_96597_(PoseStack p_96598_, List<Component> p_96599_, int p_96600_, int p_96601_) {
 -      this.m_96617_(p_96598_, Lists.transform(p_96599_, Component::m_7532_), p_96600_, p_96601_);
-+      this.renderComponentToolTip(p_96598_, p_96599_, p_96600_, p_96601_, f_96547_);
++      List<ClientTooltipComponent> components = net.minecraftforge.client.MinecraftForgeClient.gatherTooltipComponents(this.tooltipStack, p_96599_, p_96600_, p_96601_, f_96543_, f_96544_, this.tooltipFont == null ? this.f_96547_ : this.tooltipFont);
++      this.m_169383_(p_96598_, components, p_96600_, p_96601_);
```
Is there a reason these two lines are being patched in instead, rather than just having one line that makes use of the new renderComponentTooltip method that takes a Font?

diesieben07 replied:
Because that method takes a `List<FormattedText>` because of legacy reasons.

pupnewfster replied:
But it takes a list of `? extends FormattedText` so you could replace these two lines with I was pointing out with:
```java
renderComponentToolTip(p_96598_, p_96599_, p_96600_, p_96601_, this.tooltipFont == null ? this.font : this.tooltipFont);
```
and it should work just fine? Though on second thought maybe that doesn't quite make as much sense with how the font param would end up being/setting things.

diesieben07 replied:
Yeah, that'd be setting the font again which is why I didn't do this.

pupnewfster commented on patches/minecraft/net/minecraft/client/gui/screens/Screen.java.patch:
```diff
@@ -1,68 +1,104 @@
 +   }
-+   public void renderComponentToolTip(PoseStack matrixStack, List<? extends net.minecraft.network.chat.FormattedText> tooltips, int mouseX, int mouseY, Font font) {
-+      net.minecraftforge.fmlclient.gui.GuiUtils.drawHoveringText(matrixStack, tooltips, mouseX, mouseY, f_96543_, f_96544_, -1, font);
++   public void renderComponentToolTip(PoseStack poseStack, List<? extends net.minecraft.network.chat.FormattedText> tooltips, int mouseX, int mouseY, Font font) {
```
This should probably be named `renderComponentTooltip` to have consistent capitalization in the word `Tooltip` with all the other tooltip methods.

diesieben07 replied:
That would be a binary breaking change

pupnewfster replied:
Ah, good catch, reading patches of patches is a pain, and missed the fact that it was really just changing the body of the method.

pupnewfster commented on patches/minecraft/net/minecraft/client/gui/screens/Screen.java.patch:
```diff
@@ -1,68 +1,104 @@
           for(ClientTooltipComponent clienttooltipcomponent : p_169385_) {
 -            int k = clienttooltipcomponent.m_142069_(this.f_96547_);
-+            int k = clienttooltipcomponent.m_142069_(font);
++            int k = clienttooltipcomponent.m_142069_(f_96547_);
```
Is this meant to be `event#getFontRenderer()`? If not, then adding back the `this.` before the passed variable would allow for this line of the patch to be removed, though my suspicion is it is meant to get it from the event.

pupnewfster commented on patches/minecraft/net/minecraft/client/gui/screens/Screen.java.patch:
```diff
@@ -1,68 +1,104 @@
+-         m_93123_(matrix4f, bufferbuilder, j2 - 3, k2 + j + 2, j2 + i + 3, k2 + j + 3, 400, 1344798847, 1344798847);
++         net.minecraftforge.client.event.RenderTooltipEvent.Color colorEvent = net.minecraftforge.client.MinecraftForgeClient.colorTooltipEvent(this.tooltipStack, p_169384_, p_169386_, p_169387_, preEvent.getFontRenderer(), p_169385_);
++         j2 = preEvent.getX();
++         k2 = preEvent.getY();
```
I think moving these two lines above the firing of the color event would be a bit easier to read, in that it is more clear that it isn't grabbing them from the color event, and is mostly just here for reducing patch size as they are not needed earlier. Though part of me is wondering if the actual values should also be passed into the color tooltip event as well in case someone wants to make use of those values for calculating things but they ended up being changed by someone else in the pre event.

pupnewfster commented on patches/minecraft/net/minecraft/client/gui/screens/Screen.java.patch:
```diff
@@ -1,68 +1,104 @@
++         m_93123_(matrix4f, bufferbuilder, j2 - 3, k2 - 3 + 1, j2 - 3 + 1, k2 + j + 3 - 1, 400, colorEvent.getBorderStart(), colorEvent.getBorderStart());
++         m_93123_(matrix4f, bufferbuilder, j2 + i + 2, k2 - 3 + 1, j2 + i + 3, k2 + j + 3 - 1, 400, colorEvent.getBorderStart(), colorEvent.getBorderStart());
++         m_93123_(matrix4f, bufferbuilder, j2 - 3, k2 - 3, j2 + i + 3, k2 - 3 + 1, 400, colorEvent.getBorderStart(), colorEvent.getBorderStart());
++         m_93123_(matrix4f, bufferbuilder, j2 - 3, k2 + j + 2, j2 + i + 3, k2 + j + 3, 400, colorEvent.getBorderStart(), colorEvent.getBorderStart());
```
Are some of these meant to go between `getBorderStart` and `getBorderEnd`? Also given it is a gradient, would there be any benefit to allowing there to be two background colors so that people can make the background itself go between multiple colors.

pupnewfster commented on src/main/java/net/minecraftforge/fmlclient/gui/GuiUtils.java:
```diff
@@ -237,6 +237,7 @@ public static void drawTexturedModalRect(PoseStack matrixStack, int x, int y, in
      *
      * @param stack The stack for which a tooltip is about to be drawn.
      */
+    @Deprecated
```
All these methods that are being deprecated potentially should have a TODO for 1.18 to remove them.

pupnewfster commented on src/main/java/net/minecraftforge/client/event/RenderTooltipEvent.java:
```diff
@@ -45,20 +50,29 @@
 {
     @Nonnull
     protected final ItemStack stack;
+    @Deprecated
```
I feel like the things that are being deprecated in this class probably should have TODO's for 1.18 to remove them.

pupnewfster commented on src/main/java/net/minecraftforge/client/event/RenderTooltipEvent.java:
```diff
@@ -73,14 +87,21 @@ public ItemStack getStack()
      * 
-     * @return An <i>unmodifiable</i> list of strings. Use {@link ItemTooltipEvent} to modify tooltip text.
+     * @return An <i>unmodifiable</i> list of strings. Use {@link net.minecraftforge.event.entity.player.ItemTooltipEvent} to modify tooltip text.
+     * @deprecated use {@link #getComponents()}
```
If this method is deprecated as indicated by the change in javadocs the method should probably actually be annotated with `@Deprecated` as well

pupnewfster commented on src/main/java/net/minecraftforge/client/event/RenderTooltipEvent.java:
```diff
@@ -73,14 +87,21 @@ public ItemStack getStack()
     }
 
+    @Nonnull
+    public List<ClientTooltipComponent> getComponents()
```
It seems like this method should have a javadoc similar to what getLines had?

pupnewfster commented on src/main/java/net/minecraftforge/client/event/RenderTooltipEvent.java:
```diff
@@ -106,39 +127,108 @@ public int getY()
+            this.maxWidth = maxWidth;
+        }
+
+        public ItemStack getStack()
```
While this method and a bunch of the other methods being added are relatively self explanatory, it still may be worthwhile to have brief javadocs for them.

pupnewfster commented on src/main/java/net/minecraftforge/client/event/RenderTooltipEvent.java:
```diff
@@ -45,20 +50,29 @@
     protected final ItemStack stack;
+    @Deprecated
     protected final List<? extends FormattedText> lines;
     protected final PoseStack matrixStack;
```
Probably worth a TODO on this for 1.18 to rename it to be more in line with vanilla, as well as the getter for it.

pupnewfster commented on src/main/java/net/minecraftforge/client/MinecraftForgeClient.java:
```diff
@@ -176,4 +192,103 @@ public static ITextureAtlasSpriteLoader getTextureAtlasSpriteLoader(ResourceLoca
+        return gatherTooltipComponents(stack, textElements, Optional.empty(), mouseX, mouseY, screenWidth, screenHeight, font);
+    }
+
+    public static List<ClientTooltipComponent> gatherTooltipComponents(ItemStack stack, List<? extends FormattedText> textElements, Optional<TooltipComponent> itemComponent, int mouseX, int mouseY, int screenWidth, int screenHeight, Font font)
```
Is it intentional that the `mouseY` parameter is not being used?

diesieben07 replied:
Yes, it is not needed for text wrapping. I'll remove it.

pupnewfster commented on src/main/java/net/minecraftforge/client/MinecraftForgeClient.java:
```diff
@@ -176,4 +192,103 @@ public static ITextureAtlasSpriteLoader getTextureAtlasSpriteLoader(ResourceLoca
+                    .toList();
+        }
+        else
+        {
```
Given it is sort of hard to read already due to the multiple lines from streams and things, I feel like it would be easier to read if the else was just removed given the if above always returns so there isn't any case the that the if can be fallen through, and it would then allow for there to be slightly less indentation.

pupnewfster reviewed (commented):
Note: I haven't gotten a chance to actually test it yet, so I may end up having a couple more comments once I do, but for now added a bunch of comments related to what I saw while reviewing the code.

diesieben07 committed:
Fix parameter names on forge-added method

diesieben07 committed:
fix forge-added method name and param names

diesieben07 committed:
fix missing font renderer patch

diesieben07 committed:
fix color event patch

diesieben07 committed:
Fix border gradient patch missing

diesieben07 committed:
Allow using a gradient for backgrounds

diesieben07 committed:
remove unnecessary Y param for gatherTooltips

diesieben07 committed:
simplify indentation

diesieben07 committed:
fix deprecations and javadocs

diesieben07 committed:
Merge branch '1.17.x' into bug/tooltip-event-rewrite

diesieben07 committed:
Merge branch '1.17.x' into bug/tooltip-event-rewrite

# Conflicts:
#	src/test/resources/META-INF/mods.toml

diesieben07 commented:
Addressed your review @pupnewfster.

pupnewfster commented on src/main/java/net/minecraftforge/client/event/RenderTooltipEvent.java:
```diff
@@ -106,39 +133,117 @@ public int getY()
+            return tooltipElements;
+        }
+
+        public int getMaxWidth()
```
Probably worth a JavaDoc

pupnewfster commented on src/main/java/net/minecraftforge/client/event/RenderTooltipEvent.java:
```diff
@@ -106,39 +133,117 @@ public int getY()
         private int screenHeight;
         private int maxWidth;
 
+        @Deprecated
```
These deprecated methods (and other nested event classes) should also probably have a TODO for removal in 1.18 given they aren't used any more and don't serve a purpose.

sciwhiz12 replied:
Along with a `// TODO: Remove in 1.18`, it should also use the `forRemoval` value to more clearly indicate that it is deprecated for removal.

pupnewfster commented on src/main/java/net/minecraftforge/client/event/RenderTooltipEvent.java:
```diff
@@ -259,30 +372,61 @@ public PostText(@Nonnull ItemStack stack, @Nonnull List<? extends FormattedText>
+            return backgroundStart;
+        }
+
+        public int getBackgroundStart()
```
I am not really sure if it serves much purpose but would it make sense to deprecate the `getOriginalBackground` method and add two corresponding ones for the original values? This may not make sense given currently the event only really allows for one original background so the values are assumed the same. (Putting the comment here as I can't add it down below where the existing method is)

pupnewfster reviewed (commented):
Couple more comments on the code, and also a few from when I was doing some testing.
1. In the test mod, is the changes to sticks, clocks, and apples supposed to be noticeable? Given I am not really seeing anything, but I may just be blind if it is a subtle change.
2. The text wrapping seems to have some issues in regards to if it properly stops itself from going off the screen or not. 
![2021-09-11_09 34 28](https://user-images.githubusercontent.com/1203288/132950140-9fb2aeae-3bf2-415f-b343-c328e9086d7b.png)

diesieben07 commented:
> In the test mod, is the changes to sticks, clocks, and apples supposed to be noticeable? Given I am not really seeing anything, but I may just be blind if it is a subtle change.

The apple demonstrates custom background and border colors:
![2021-09-12_16 23 44](https://user-images.githubusercontent.com/1915984/132991416-8e51506f-f2e0-45af-b1cb-c42d02e4320e.png)

The stick demonstrates adding custom tooltips via the event:
![2021-09-12_16 27 12](https://user-images.githubusercontent.com/1915984/132991534-febde8d3-32cb-4d02-98ac-32071d0a874a.png)

The clock demonstrates a custom maximum width:
![2021-09-12_16 28 00](https://user-images.githubusercontent.com/1915984/132991550-a3fdd359-1bab-4f13-9380-69241373fec6.png)

diesieben07 committed:
fix tooltip offset from the mouse

diesieben07 committed:
Fix missing tooltip flipping on oversize

diesieben07 committed:
Add javadocs for getMaxWidth

diesieben07 committed:
More deprecation updates

diesieben07 committed:
Add separate "original" methods for start and end of background

diesieben07 commented:
Fixed the text wrapping and position bugs, fixed the missing deprecation comments and javadocs.

pupnewfster commented on src/main/java/net/minecraftforge/client/event/RenderTooltipEvent.java:
```diff
@@ -106,39 +133,120 @@ public int getY()
         public Pre(@Nonnull ItemStack stack, @Nonnull List<? extends FormattedText> lines, PoseStack matrixStack, int x, int y, int screenWidth, int screenHeight, int maxWidth, @Nonnull Font fr)
         {
-            super(stack, lines, matrixStack, x, y, fr);
+            this(stack, matrixStack, x, y, screenWidth, screenHeight, fr, List.of());
```
While it is a bit more of a mess, I feel like this should more or less be a copy of the new Pre constructor instead of calling this, so that it can then pass `lines` to the now deprecated `RenderTooltipEvent` super constructor, in case any mods are currently listening to the events and any other mods are still using/decide to use the GuiUtils methods for some reason.

pupnewfster commented on src/main/java/net/minecraftforge/client/event/RenderTooltipEvent.java:
```diff
@@ -259,30 +375,61 @@ public PostText(@Nonnull ItemStack stack, @Nonnull List<? extends FormattedText>
+        public Color(@Nonnull ItemStack stack, @Nonnull List<? extends FormattedText> textLines, PoseStack matrixStack, int x, int y, @Nonnull Font fr, int background, int borderStart, int borderEnd)
         {
-            super(stack, textLines, matrixStack, x, y, fr);
+            this(stack, matrixStack, x, y, fr, background, borderStart, borderEnd, List.of());
```
While it is a bit more of a mess, I feel like this should more or less be a copy of the new Color constructor instead of calling this, so that it can then pass `lines` to the now deprecated `RenderTooltipEvent` super constructor, in case any mods are currently listening to the events and any other mods are still using/decide to use the GuiUtils methods for some reason.

pupnewfster commented on src/main/java/net/minecraftforge/client/event/RenderTooltipEvent.java:
```diff
@@ -45,20 +50,29 @@
+    @Deprecated(forRemoval = true) // TODO: remove in 1.18
     public RenderTooltipEvent(@Nonnull ItemStack stack, @Nonnull List<? extends FormattedText> lines, PoseStack matrixStack, int x, int y, @Nonnull Font fr)
+    {
+        this(stack, matrixStack, x, y, fr, List.of());
```
While it is a bit more of a mess, I feel like this should more or less be a copy of the new RenderTooltipEvent constructor instead of calling this, which would allow the lines parameter to be properly assigned to the lines field. This way while usage of this is deprecated it at least has proper values for the deprecated variant.

pupnewfster reviewed (commented):
Looks good overall, but when I was double checking things noticed that there are currently a couple tiny binary breaks in regards to the now deprecated `RenderTooltipEvent.lines` field not being assigned when using the deprecated constructors that takes a list of lines instead of components.

diesieben07 committed:
Improve backwards compat of old constructors

diesieben07 committed:
Improve backwards compat of old constructors

tterrag1098 commented:
Haven't had time to dig into details, but the broad strokes looks like a faithful port of the original structure. LGTM.

diesieben07 committed:
Merge branch '1.17.x' into bug/tooltip-event-rewrite

# Conflicts:
#	src/test/resources/META-INF/mods.toml

Mysterious-Dev commented:
This PR is still relevant ?

diesieben07 committed:
Fix #8124

diesieben07 committed:
Merge branch '1.17.x' into bug/tooltip-event-rewrite

# Conflicts:
#	patches/minecraft/net/minecraft/client/gui/screens/Screen.java.patch
#	src/test/resources/META-INF/mods.toml

diesieben07 committed:
- move event call hooks to ForgeHooksClient
- Streamline Font override handling
- fix tooltip stack missing from events on creative search tab

diesieben07 commented:
Updated again:

- Resolved merge conflicts
- Moved event call hooks to `ForgeHooksClient`
- Fixed tooltip events firing without `ItemStack` info on creative menu search tab
- Streamlined font overriding handling via helper method in `ForgeHooksClient`

diesieben07 committed:
Merge branch '1.17.x' into bug/tooltip-event-rewrite

# Conflicts:
#	src/test/resources/META-INF/mods.toml

diesieben07 commented:
Resolved merge conflicts again

diesieben07 committed:
update deprecated annotations

diesieben07 committed:
update deprecated annotations
